### PR TITLE
feat: comparison of assemblers

### PR DESCRIPTION
**Pull request by alethomas**
New advanced comparison of all assemblers with measures and figure for the publication

### CODE REVIEW

File: workflow/rules/benchmarking.smk [hunk 1]
```diff
@@ -204,3 +204,242 @@ rule plot_strain_call_error:
...
+#     conda:
+#         "../envs/abyss.yaml"
+#     shell:
+#         """
+#         if [ -d "{params.outdir}" ]; then rm -Rf {params.outdir}; fi
+#         (mkdir -p {params.outdir} && cd {params.outdir}
+#         abyss-pe np={threads} name={wildcards.sample} k=96 in='../../../../../{input.fastq1} ../../../../../{input.fastq2}'
+#         cd ../../../../../ && mv {params.outdir}/{wildcards.sample}-contigs.fa {output} ) > {log} 2>&1
+#     """
+
```
Comment (thomasbtf) [inline workflow/rules/benchmarking.smk:267]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                    'Assembler': assembler, 
+                    'bp': max(len(contig.sequence) for contig in infile_final),
+                    'Contig': 'final',
+                    },
+                    ignore_index=True
+                )
+    return data
+                # data.loc[sample, "Assembler"] = assembler
+                # data.loc[sample, "Initial Contig (bp)"] = max(len(contig.sequence) for contig in infile_init)
+                # data.loc[sample, "Final Contig (bp)"] = max(len(contig.sequence) for contig in infile_final)
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:46]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+for assembler in dict_init:
+    data=register_contig_lengths(dict_init[assembler], dict_final[assembler], assembler, data)
+
+# data.reset_index(inplace=True)
+
+
+# init = alt.Chart(data).mark_boxplot(color='#57A44C').encode(
+#     alt.Y('Initial Contig (bp)',
+#           axis=alt.Axis(title='# bp')),
+# )
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:59]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+# line = base.mark_line(stroke='#5276A7', interpolate='monotone').encode(
+#     alt.Y('Initial Contig (bp) megahit',
+#           axis=alt.Axis(title='Precipitation (inches)', titleColor='#5276A7'))
+
+plot1.save(snakemake.output[0])
+plot2.save(snakemake.output[0])
+# (plot2+plot3).save(snakemake.output[0])
+# with open(snakemake.output[0], "w") as outfile:
+
+#     pass
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:112]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+).configure_view(
+    stroke=None
+)
+
+print(data)
+# line = base.mark_line(stroke='#5276A7', interpolate='monotone').encode(
+#     alt.Y('Initial Contig (bp) megahit',
+#           axis=alt.Axis(title='Precipitation (inches)', titleColor='#5276A7'))
+
+plot1.save(snakemake.output[0])
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:107]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+).transform_calculate(
+    # Generate Gaussian jitter with a Box-Muller transform
+    jitter='sqrt(-2*log(random()))*cos(2*PI*random())'
+# ).configure_facet(
+#     spacing=0
+).configure_view(
+    stroke=None
+)
+
+print(data)
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:102]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        # ),
+    ),
+).transform_calculate(
+    # Generate Gaussian jitter with a Box-Muller transform
+    jitter='sqrt(-2*log(random()))*cos(2*PI*random())'
+# ).configure_facet(
+#     spacing=0
+).configure_view(
+    stroke=None
+)
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:100]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+#     alt.Y('Initial Contig (bp)',
+#           axis=alt.Axis(title='# bp')),
+# )
+plot1 = alt.Chart(data).mark_boxplot(color='#08A4C2').encode(
+    alt.Y('bp'),
+    color='Contig',
+    x=('Assembler')
+)
+
+plot2 = alt.Chart(data).mark_boxplot().encode(
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:66]:
```suggestion
boxplot = alt.Chart(data).mark_boxplot().encode(
```

File: workflow/scripts/plot-assembly-comparison.py (hunk 5 above)
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:107]:
```suggestion
boxplot.save(snakemake.output[0])
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+# init = alt.Chart(data).mark_boxplot(color='#57A44C').encode(
+#     alt.Y('Initial Contig (bp)',
+#           axis=alt.Axis(title='# bp')),
+# )
+plot1 = alt.Chart(data).mark_boxplot(color='#08A4C2').encode(
+    alt.Y('bp'),
+    color='Contig',
+    x=('Assembler')
+)
+
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:65]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
+import os
+import sys
+import pandas as pd
+from pandas.core.indexes.base import Index
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:4]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
+import os
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:1]:
```suggestion
```

**Review by thomasbtf — CHANGES_REQUESTED** [review]
Nice 👍 
Maybe run black on the script

File: workflow/Snakefile [hunk 12]
```diff
@@ -138,6 +138,14 @@ checkpoint all:
         ),
 
 
+rule compare_assemlbers:
```
Comment (johanneskoester) [inline workflow/Snakefile:141]:
```suggestion
rule compare_assemblers:
```

File: workflow/Snakefile [hunk 13]
```diff
@@ -138,6 +138,14 @@ checkpoint all:
...
+    input:
+        expand(
+            "results/{date}/plots/all_assemblies.svg",
```
Comment (johanneskoester) [inline workflow/Snakefile:144]: Maybe better call this `plots/assembler-comparison.svg`?

File: workflow/rules/benchmarking.smk [hunk 14]
```diff
@@ -206,6 +206,271 @@ rule plot_strain_call_error:
...
+        "mv {params.outdir}/contigs.fasta {output.contigs}) > {log} 2>&1"
+
+
+rule assembly_comparison_rnaviralspades:
+    input:
+        fastq1=lambda wildcards: get_reads_after_qc(wildcards, read="1"),
+        fastq2=lambda wildcards: get_reads_after_qc(wildcards, read="2"),
+    output:
+        contigs=(
+            "results/{date}/assembly/{sample}/rnaviralspades/{sample}.contigs.fasta"
```
Comment (johanneskoester) [inline workflow/rules/benchmarking.smk:366]: all the different spades rules look exactly the same. Just have one, and turn the type into a wildcard
```suggestion
            "results/{date}/assembly/{sample}/{spadesflavor}/{sample}.contigs.fasta"
```

Then, in the shell command start with
```
shell:
    "{wildcards.spadesflavor}.py ..."
```

File: workflow/rules/benchmarking.smk [hunk 15]
```diff
@@ -206,6 +206,271 @@ rule plot_strain_call_error:
...
+    log:
+        "logs/{date}/megahit-meta-sensitive/{sample}.log",
+    params:
+        outdir=lambda w, output: os.path.dirname(output[0]),
+    threads: 8
+    conda:
+        "../envs/megahit.yaml"
+    shell:
+        "(megahit -1 {input.fastq1} -2 {input.fastq2} --presets meta-sensitive --out-dir {params.outdir} -f && "
+        "mv {params.outdir}/final.contigs.fa {output.contigs} ) > {log} 2>&1"
```
Comment (johanneskoester) [inline workflow/rules/benchmarking.smk:262]: All the megahit rules are basically the same. Just have one, add a wildcard for the preset (e.g. `{megahit_preset}` and set the parameters accordingly.

Something like this:
```suggestion
rule assembly_comparison_megahit_meta_sensitive:
    input:
        fastq1=lambda wildcards: get_reads_after_qc(wildcards, read="1"),
        fastq2=lambda wildcards: get_reads_after_qc(wildcards, read="2"),
    output:
        contigs="results/{date}/assembly/{sample}/megahit-{preset}/{sample}.contigs.fasta",
    log:
        "logs/{date}/megahit-meta-sensitive/{sample}.log",
    params:
        outdir=lambda w, output: os.path.dirname(output[0]),
    threads: 8
    conda:
        "../envs/megahit.yaml"
    shell:
        "(megahit -1 {input.fastq1} -2 {input.fastq2} --presets {wildcards.preset} --out-dir {params.outdir} -f && "
        "mv {params.outdir}/final.contigs.fa {output.contigs} ) > {log} 2>&1"
```

**Review by johanneskoester — CHANGES_REQUESTED** [review]

File: workflow/scripts/plot-assembly-comparison.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        amplicon_state,
+        data,
+    )
+    data = register_lengths(
+        sample,
+        [x for x in snakemake.input.quast if sample in x],
+        "N50",
+        amplicon_state,
+        data,
+    )
```
Comment (johanneskoester) [inline workflow/scripts/plot-assembly-comparison.py:68]: code duplication, turn into a function and run 3 times with different parameters instead.

**Review by johanneskoester — CHANGES_REQUESTED** [review]

File: config/config.yaml [hunk 17]
```diff
@@ -124,3 +124,15 @@ voc:
     - E484K
     - K417N
     - H69_V70del
+
+# assemblers used for assembly comparison
+assemblers_for_comparison:
+  - megahit-std
```
Comment (thomasbtf) [inline config/config.yaml:130]: Maybe we should include some of the assemblers in the .tests config and include the assembler comparison in GitHub Actions.

File: workflow/rules/benchmarking.smk [hunk 18]
```diff
@@ -206,6 +206,186 @@ rule plot_strain_call_error:
...
+        "logs/{date}/{spadesflavor}/{sample}.log",
+    threads: 8
+    conda:
+        "../envs/spades.yaml"
+    shell:
+        "({wildcards.spadesflavor}.py -1 {input.fastq1} -2 {input.fastq2} -o {params.outdir} -t {threads} && "
+        "mv {params.outdir}/contigs.fasta {output.contigs}) > {log} 2>&1"
+
+
+rule order_contigs_assembly_comparison:
```
Comment (thomasbtf) [inline workflow/rules/benchmarking.smk:295]: [Rule Inheritance](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#rule-inheritance) could be a better choice here than duplicate the rules. Even better would be the usage of the original rules.

File: workflow/rules/benchmarking.smk [hunk 19]
```diff
@@ -206,6 +206,186 @@ rule plot_strain_call_error:
...
+    threads: 8
+    conda:
+        "../envs/ragoo.yaml"
+    shell:  # currently there is no conda package for mac available. Manuell download via https://github.com/malonge/RaGOO
+        "(cd {params.outdir} && "
+        "ragoo.py ../../../../../{input.contigs} ../../../../../{input.reference} && "
+        "cd ../../../../../ && mv {params.outdir}/ragoo_output/ragoo.fasta {output}) > {log} 2>&1"
+
+
+rule filter_chr0_assembly_comparison:
```
Comment (thomasbtf) [inline workflow/rules/benchmarking.smk:314]: See comment about [Rule Inheritance](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#rule-inheritance) above

File: workflow/rules/benchmarking.smk [hunk 20]
```diff
@@ -206,6 +206,186 @@ rule plot_strain_call_error:
...
+        "logs/{date}/ragoo/{assembler}/{sample}_cleaned.log",
+    params:
+        sample=lambda wildcards: wildcards.sample,
+    conda:
+        "../envs/python.yaml"
+    script:
+        "../scripts/ragoo-remove-chr0.py"
+
+
+rule align_contigs_assembly_comparison:
```
Comment (thomasbtf) [inline workflow/rules/benchmarking.smk:329]: See comment about [Rule Inheritance](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#rule-inheritance) above

File: workflow/rules/benchmarking.smk [hunk 21]
```diff
@@ -206,6 +206,186 @@ rule plot_strain_call_error:
...
+        "results/{date}/assembly/{sample}/{assembler}/main_{sample}.bam",
+    log:
+        "results/{date}/assembly/{sample}/{assembler}/main_{sample}.log",
+    conda:
+        "../envs/minimap2.yaml"
+    shell:
+        "minimap2 -ax asm5 {input.target} {input.query} -o {output} 2> {log}"
+
+
+rule quast_assembly_comparison:
```
Comment (thomasbtf) [inline workflow/rules/benchmarking.smk:343]: See comment about [Rule Inheritance](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#rule-inheritance) above

File: workflow/scripts/plot-assembly-comparison.py [hunk 22]
```diff
@@ -0,0 +1,154 @@
+import sys
+from altair.vegalite.v4.schema.channels import Opacity
+import pandas as pd
+import pysam
+import altair as alt
+
+sys.stderr = open(snakemake.log[0], "w")
+
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:8]:
```suggestion
import sys

sys.stderr = open(snakemake.log[0], "w")

import pandas as pd
import pysam
import altair as alt

```

File: workflow/scripts/plot-assembly-comparison.py [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+            )
+    return data
+
+
+for sample, amplicon_state in zip(
+    snakemake.params.samples, snakemake.params.amplicon_state
+):
+
+    def load(inputfile, label=None):
+        return register_lengths(
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:49]: Function is defined inside a `for` statement

File: workflow/scripts/plot-assembly-comparison.py [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+            axis=alt.Axis(tickCount=5),
+        ),
+        x=alt.X("State", title=None, sort=["initial", "scaffolded", "N50"]),
+        color=alt.Color("Assembler", scale=alt.Scale(scheme="turbo"), legend=None),
+    )
+    .properties(height=height, width=width)
+)
+
+combined_bp = plot_bp.mark_point(opacity=0.5, filled=True)
+# combined_bp = plot_bp.mark_boxplot(opacity=0.8, color='black')
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:99]:
```suggestion
```

File: workflow/scripts/plot-assembly-comparison.py [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+
+plot_bp = (
+    alt.Chart(data)
+    .encode(
+        y=alt.Y(
+            "length (bp)",
+            scale=alt.Scale(domain=[0, 35000], clamp=True),
+            axis=alt.Axis(tickCount=5),
+        ),
+        x=alt.X("State", title=None, sort=["initial", "scaffolded", "N50"]),
```
Comment (thomasbtf) [inline workflow/scripts/plot-assembly-comparison.py:92]: We should rename the contig state in something more meaningful for outsiders and put the new 50 in the first second place, maybe like:
`N50, Largest Contig, Scaffold`

File: workflow/rules/common.smk [hunk 26]
```diff
@@ -668,6 +679,30 @@ def get_assemblies_for_submission(wildcards, agg_type):
...
+            sample=get_samples_for_date(wildcards.date),
+            **kwargs,
+        )
+
+    return inner
+
+
+def expand_samples_for_date_assembler(paths, **kwargs):
+    return expand_samples_by_func(
+        paths, get_samples_for_date, assembler=config.get("assemblers_for_comparison")
```
Comment (thomasbtf) [inline workflow/rules/common.smk:695]:
```suggestion
        paths, get_samples_for_date, assembler=config["assemblers_for_comparison"]
```

File: workflow/rules/common.smk [hunk 27]
```diff
@@ -555,6 +555,17 @@ def get_samples_for_date_amplicon(date):
...
 
+def get_list_of_amplicon_states(date):
+    samples = get_samples_for_date(date)
+    amplicon_states = []
+    for sample in samples:
+        if is_amplicon_data(sample):
+            amplicon_states.append(True)
+        else:
+            amplicon_states.append(False)
+    return amplicon_states
```
Comment (thomasbtf) [inline workflow/rules/common.smk:566]:
```suggestion
return [True if is_amplicon_data(sample) else  False for sample in samples ]
```

**Review by thomasbtf — CHANGES_REQUESTED** [review]
Again, love the plot 😊 Here some remarks

File: workflow/scripts/plot-assembly-comparison.py (hunk 23 above)
Comment (alethomas) [reply workflow/scripts/plot-assembly-comparison.py:49]: That's a closure, suggested by @johanneskoester

File: workflow/scripts/plot-assembly-comparison.py (hunk 25 above)
Comment (alethomas) [reply workflow/scripts/plot-assembly-comparison.py:92]: I agree N50 first makes more sense, as it is related with the initial contigs

**Comment by alethomas** [comment]
assembler-comparison
assembler-comparison_genome_fraction

File: workflow/scripts/plot-assembly-comparison.py [hunk 28]
```diff
@@ -0,0 +1,179 @@
+import sys
+from altair.vegalite.v4.schema.channels import Opacity
+from altair.vegalite.v4.schema.core import TitleOrient
```
Comment (alethomas) [inline workflow/scripts/plot-assembly-comparison.py:3]: remov
```suggestion
```

File: workflow/rules/common.smk [hunk 29]
```diff
@@ -344,41 +344,38 @@ def get_min_coverage(wildcards):
...
 
 
-def get_contigs(wildcards):
-    if is_amplicon_data(wildcards.sample):
-        pattern = (
-            "results/{date}/assembly/metaspades/{sample}/{sample}.contigs.fasta",
-        )
+def return_assembler(sample):
+    if is_amplicon_data(sample):
+        return config.get("assembler_amplicon")
```
Comment (alethomas) [inline workflow/rules/common.smk:349]:
```suggestion
        return config["assembler_amplicon"]
```

File: .tests/config/config.yaml [hunk 30]
```diff
@@ -128,4 +128,22 @@ voc:
...
+  - metaspades
+  - coronaspades
+  - spades
+  - rnaviralspades
+
+# assemblers used for sequencing method 
+assembler_shotgun:
+  megahit-std
+assembler_amplicon:
+  metaspades
```
Comment (alethomas) [inline .tests/config/config.yaml:149]: also change all instances
```suggestion
assemblers_used:
 - shotgun: megahit-std
 - amplicon: metaspades
```

**Review by thomasbtf — APPROVED** [review]

**Pull request merged**